astropy/units/quantity_helper/function_helpers.py

```diff
@@ -572,6 +573,20 @@ def aslayoutarray_helper(a, dtype=None):
         a = _as_quantity(a).value
     return (a, dtype), {}, out_unit, None
 
+@function_helper(helps={np.frombuffer, np.fromfile, np.fromiter, np.fromstring})
+def fromrawdata_helper(*args, **kwargs):
+    return args, kwargs, UNIT_FROM_LIKE_ARG, None
+
+@function_helper
+def fromfunction(function, shape, *, dtype=float, **kwargs):
+    zero_arg = (0,) * len(shape)
```

Comment:
I guess this is one function where having it dispatched would be easier... But I like your solution, except I think it should be `zero_arg = np.zeros(len(shape), dtype)`, so that the functions gets passed the right dtype (`*zero_arg` will still work).

Reply:
nice catch !